### ISSUE TITLE
DeveloperExceptionPageMiddleware always returns HTML, sometimes JSON would be better

### ISSUE THREAD

**Issue by aspnet-hello** [aspnet/AspNetCore#2590]
_From @tjrobinson on Friday, February 3, 2017 3:54:20 AM_

We're using the DeveloperExceptionPageMiddleware and it would be really useful if it could be changed/configured so that AJAX requests to APIs, e.g. from an Angular application were given a JSON response back instead of HTML. This would make the response much more readable when using F12 tools or testing the API through other non-browser tools like Postman.

I hope that's enough information to make sense, let me know if not.

There's a similar problem discussed here: http://stackoverflow.com/questions/35245893/mvc-6-webapi-returning-html-error-page-instead-of-json-version-of-exception-obje

Thanks!

_Copied from original issue: aspnet/Diagnostics#346_

**Comment by aspnet-hello** [comment]
_From @davidfowl on Sunday, August 6, 2017 12:15:32 PM_

Feels like we could do this as part of the new API work @rynowak.

/cc @Eilon

**Comment by aspnet-hello** [comment]
_From @rynowak on Sunday, August 6, 2017 9:18:48 PM_

https://tools.ietf.org/html/rfc7807

And then you want formatters and get bit by the layering bug 😆

**Comment by mkArtakMSFT** [comment]
Sounds like the ask here is mainly for developers. Postman, for example, has the "preview" option which allows to see the HTML responses like in browser.

**Comment by mkArtakMSFT** [comment]
Thank you for your feature request. We'll consider this feature during the next release planning period and update the status of this issue accordingly.

**Comment by Eilon** [comment]
I've been thinking about this space and started some prototypes, so self-assigning for now.

**Comment by PureKrome** [comment]
Thanks heaps @Eilon !

**Comment by khellang** [comment]
See [this Twitter thread](https://twitter.com/khellang/status/1009561214857416704) if you want something for APIs 😄

**Comment by rynowak** [comment]
So we've done extensible handlers for DeveloperExceptionPage, and we've done a fallback to plaintext for non-html-clients (ie, not browsers).

That feels like enough for a default experience here. Since this is for dev-time, we don't think JSON is going to end up being more human-readable than plaintext. 

Feel free to post your feedback if you think have a JSON version of the **development exception page** is vital. We have some other plans for improved error handling in general for APIs that relate to production that are tracked elsewhere.

**Comment by PureKrome** [comment]
> Feel free to post your feedback if you think have a JSON version of the development exception page is vital.

@rynowak where do we post feedback about this statement? here in this issue?

**Comment by rynowak** [comment]
Sure. If you're interested in this

**Comment by PureKrome** [comment]
Thanks for listening @rynowak .

My thoughts in this are pretty weak - plain text solution basically solves it all :) But I'll just jot down my thoughts cause this is the opportunity. Maybe someone else might get an 'a-ha' moment, based of this. 

But having plain text is soo awesome! It will be usable :)

---

So this is all about *web api*'s -> which usually are responses sent back as `JSON`-or-`XML` (either hardcoded by the developer(s) or dynamic via `Connection Negotiation` (which again is usually `JSON/XML` again .. because this is an API.

So with this in mind, I thought it might be nice to make it easier to parse the error message(s) on the client side. Plain text is 💯 waaaay better than before (cheers! 🍻 ) but I sometimes like to display various parts of the error message to the browser client.

So the scenario is this:
- Test with Postman : errors can be nice text (as is now the option) or json/xml (more verbose, doesn't gain anything really) (rinse/repeat a bazillion times....)

then finally

- Start testing my web app (which does Ajax/API calls to my backend api) and some errors occur so we could display the errors a bit nicer if we could split out the key/values. But having plain text is sooo much better and usable than before.

I also understand that adding the extra complexity of JSON or XML output might pull in other dependencies, etc... so this could be a crap thing. Maybe have an option for people to extend this themselves, easily, if they really want  .. similar to how logging only has a few targets/sinks but can be extended easily for others. Prolly not worth the effort?

Anways, that's my 2cents. Cheers for the plain text change! appreciated!!!!

**Comment by ggirard07** [comment]
Not sure this is a real good case either but simply because our clients always parse response from the API as JSON as this is the only content type we use/support. Having a plain text response is going to throw exceptions on the client side that will only add noise the the debug experience...

**Comment by khellang** [comment]
You know... There are alternatives to this linked further up in this issue 😂😜

**Comment by rynowak** [comment]
Thanks for the feedback - some high level thoughts on this. 

We divide error handling into *Developer* error handing and *Production* error handling. That's why you use see:

```C#
if (env.IsDevelopment())
{
    app.UseDeveloperExceptionPage();
}
else
{
    app.UseErrorHandler("/some/path"); 
}
```

The developer exception page pretty prints a bunch of stuff about the request, but the *main event* is giving you the exception details. This middleware is needed because we don't want to anything built into the stack by default that does this. If it's not built in, you can remove it or replace it because it's just a line of code in your startup. The emphasis here is on being informative by default since it's for a development scenario.

The error handler middleware runs *your code* to do what you want for error handling. We have another work item planned for 3.0 to make this better by default for APIs https://github.com/aspnet/AspNetCore/issues/8539 - but the emphasis here is that you as the application owner need to make a decision about what it means to report error an error.

In the past we had some pretty complicated logic that tried to figure out if we would show you the exception details or not.  We made a conscious decision to separate error handling for development and error handling for production scenarios using code to make this clear and put the application developer in control.

-----

Responding to your comments:

> Test with Postman : errors can be nice text (as is now the option) or json/xml (more verbose, doesn't gain anything really) (rinse/repeat a bazillion times....)

This is what expected to get better. I would argue that making this experience based on JSON would be *worse* than plain text. You're going to get escaped newlines in stack trace etc.

> Start testing my web app (which does Ajax/API calls to my backend api) and some errors occur so we could display the errors a bit nicer if we could split out the key/values. But having plain text is sooo much better and usable than before.

My question here is do you actually want to see exception details from the server in your browser console (development error handling), or do you want to see what's going to come from your production error handling path?

The idea that someone is going to write code to *parse* data that we return from the DEVELOPMENT error page makes me a little nervous. This inevitably leads to people putting the DEVELOPMENT error page in production so that they can keep using this, which we'd strongly recommend against.

These factors lead me to thinking that returning JSON with the exception details is a bad default. I think it's totally fine if you want to use extensibility (or the example that @khellang ) linked to do this for your apps. You chose that after all. I haven't seen anything yet that makes me to enable this as a default.

**Comment by khellang** [comment]
My middleware (that always returns HTTP problem details as JSON) will include exception details when `IsDevelopment` is `true` by default. But it will still be there in production.

My client-side apps that consume these APIs are then guaranteed a particular response format and use a React component (kinda like smooth-code/error-overlay-webpack-plugin) that will format these errors nicely for a really smooth dev experience.

So yeah, I definitely think this is a valid story. But don't really need it built-in or on by default as I've already scratched my own itch 😉

**Comment by PureKrome** [comment]
@khellang when I was writing those phew paragraphs from a comment (a few slots above this comment) I was thinking of your lovely `ProblemDetails` library and nearly deleted my entire reply because I think that's the best solution IMO.

I've also mentioned in another GH issue _somewhere_ about how, IMO, all-and-any errors in an ASP.NET Core webapp should return a PD for all responses .. and this functionality should be baked into the framework.

But these days, you're library is a god send and, IMO, should be the default error handling.

**Comment by syndicatedshannon** [comment]
**TLDR:** Isn't this actually better solved through enhancing `UseExceptionHandler()` and `UseStatusCodePagesWithReExecute()`?

First, sorry to weigh in, in a rather pattern-opinionated manner, after some work is already underway or complete. In my team's defence, exception handling patterns have changed so frequently in ASP that we couldn't dig into this until now.

The primary issue `UseDeveloperPages` addresses is balancing development ease with security of sensitive data. I agree with the premise of this issue that this is relevant to both WebAPI/JSON and MVC/HTML.

I disagree with the [premise here](https://github.com/microsoft/aspnet-api-versioning/issues/221#issuecomment-374670503) that the format should be differentiated by route. For appropriate SoC it is typically a question of the requested format (i.e. the HTTP Accept header), not the route. We could still evaluate the Accept header in `MapWhen()`, but... *shudders at the exponential code explosion*

This issue has been around for some time in different forms, seen 3 years ago in https://github.com/aspnet/Mvc/issues/5130, https://github.com/aspnet/Diagnostics/issues/346, and https://github.com/dotnet/templating/issues/949. Other occurrences surely exist.

I agree with @dotnetchris 's [premise](https://github.com/aspnet/AspNetCore/issues/2590#issuecomment-432385334) that we generally want the same sensitive information hidden or exposed, regardless of whether we are returning JSON or HTML. We also typically map our custom Exception hierarchy into status codes that apply regardless of response format.

As mentioned frequently in these related threads, for a JSON error result to be useful, we require a fixed data structure. Without ASP.NET forcing a JSON format (while acknowledging [RFC7807](https://tools.ietf.org/html/rfc7807) exists), don't we have to assume the developer is creating a custom response?

I have another concern with `UseDeveloperExceptionPages()`, in that it encourages code to go to production without testing. Instead of switching code blocks between development and production, we should encourage minimal augmentation in development.

Lastly, I personally wish all custom error handling in ASP.NET would make it much easier to hand off more information; at a minimum a unique problem ID (e.g. GUID).

**Comment by syndicatedshannon** [comment]
Ultimately, what I personally would prefer to be able to accomplish effortlessly is:
1. Transform the exception. Do this the same way regardless of the environment or requested format.
2. Add a unique problem ID. This could potentially be included in the transformed exception.
3. Add other context, especially Identity key. This could potentially be included in the transformed exception.
4. Log the Exception and the augmented information.
5. Apply a StatusCode conversion. This could potentially be included in the transformed exception.
6. Capture/forward the IHostingEnvironment (through a either default parameter or lambda) status as a flag to display sensitive debugging information.
7. Diverge response format at the last possible moment to reduce code duplication and improve development code coverage, as is the practice in most ASP.NET implementations.
8. Include the above-identified information to the appropriate HTML or JSON renderer.
9. Continue to offer default implementation of HTML exception renderers, but also offer a default  RFC7807 rendered for JSON.
10. Include the augmented information in the rendered template 
11. Allow override of both JSON and HTML renderers (similar to `UseExceptionHandler()` today).

**Comment by davidfowl** [comment]
I don’t think we’d build something that supports all of those features. It does sound like something that could be written in a custom middleware though. I’d suggest doing so and we’ll add any supporting platform features you’d need to achieve that.

**Comment by syndicatedshannon** [comment]
@davidfowl I agree. I wouldn't expect the platform to completely implement my specific requirements.

I should note that we already do all of the above in middleware. We did have to re-implement blocks which I would not expect. So when I think about platform features I expect, I mainly look to what is already in the ASP.NET Core library; if I have to copy multiple significant blocks mostly unedited from the repo (e.g. [StatusCodePages](https://github.com/aspnet/AspNetCore/blob/0859105e86d5648134450a6a236906110f6d5bea/src/Middleware/Diagnostics/src/StatusCodePage)) to make use of it, my instinct is to ask for it to be exposed in a new way.

I'll give it some thought and follow up within a day.

**Comment by khellang** [comment]
@syndicatedshannon See https://github.com/aspnet/AspNetCore/issues/2590#issuecomment-399365000. It does pretty much all you mentioned 😊

**Comment by syndicatedshannon** [comment]
@khellang Very nice code, thank you. Unfortunately it doesn't comply well with our pre-existing opinions expressed in our code. I'm going to think more about this, perhaps my own opinions will be more flexible given a few days. In the meantime, are you available for a chat?

**Comment by syndicatedshannon** [comment]
@davidfowl So, did more digging.

Much of the core library code to accomplish [the workflow I mentioned](https://github.com/aspnet/AspNetCore/issues/2590#issuecomment-540310292) is not reusable:
 - [UseErrorHandler](https://github.com/aspnet/AspNetCore/blob/master/src/Middleware/Diagnostics/src/ExceptionHandler/ExceptionHandlerMiddleware.cs#L90), including the following knowledge:
   - application of ExceptionDispatchInfo
   - code to re-execute pipeline
   - pass Exception on DI stack as IExceptionHandlerPathFeature
 - [UseDeveloperExceptionPage](https://github.com/aspnet/AspNetCore/blob/10863cfaead78e02eb788611fc1f64a889db6941/src/Middleware/Diagnostics/src/DeveloperExceptionPage/DeveloperExceptionPageMiddleware.cs#L130) including this work:
   - capture/display headers and other HTTP context
   - parse/format both compile and runtime exceptions
   - another process to re-execute request
 - [UseStatusCodePagesWithReexecute](https://github.com/aspnet/AspNetCore/blob/10863cfaead78e02eb788611fc1f64a889db6941/src/Middleware/Diagnostics/src/StatusCodePage/StatusCodePagesExtensions.cs#L164) includes knowledge such as:
   - something apparently about and IE document size defect
   - new handling for endpoints and clearing cache
   - rendering reason phrases
   - another process to re-execute request

None of that is accessible. I also see work on various threads going towards rendering JSON problem details, and if it follows the same patterns may also be inaccessible.

Regarding new features, the main new functionality I'm seeking is something we see accomplished in a different way in each of the above: the ability to re-execute a request to another endpoint, passing a model (in this case, the ExceptionDispatchInfo, the original HTTP contextual information such as the caller's Identity, and the unique exception identity assigned at logging time).

**Comment by syndicatedshannon** [comment]
Many of the things that happen in `UseErrorHandler`, and other ASP.NET Core handlers, we want to happen regardless of formatter.
 - Discrimination of sensitive and non-sensitive information.
 - Status code assignment.
 - Mapping of proprietary human-readable messages.
 - ExceptionDispatch pattern.
 - TraceID or ExceptionId association (ideally something that can be provided to end-user).

We also probably want to avoid branching middleware to a separate pipeline in development, because it can cause non-integration defect discovery to be deferred to integration.

I know this is an ugly/incomplete location for this conversation; I'm still thinking it through and promised I'd get back to you.

**Comment by syndicatedshannon** [comment]
@khellang Getting back to you on the library you recommended; it's very nice, but AFAICT it only handles the last step of what I just described above, requires early branching, and doesn't provide an easy means to receive the pre-determined information. I agree it does integrate similarly as an additional middleware layer with the existing error handling. For example, it requires re-statement of StatusCode maps via `options.Map<NotImplementedException>(ex => new ExceptionProblemDetails(ex, StatusCodes.Status501NotImplemented));`

**Comment by syndicatedshannon** [comment]
> What is that step exactly? It's hard to tell which 

Agreed :) I was editing the referenced comment quite a bit. I think I was referring to RFC ProblemDetails at the time.

> There's no branching involved in the middleware.

Agreed. The branching has to occur earlier, prior to your middleware, cutting various steps entirely, to handle development scenarios, etc. Some code somewhere has to branch, to respond to format (HTML/JSON) and sensitive display (IsDevelopment). IMO, ideally the branching, especially when it will never see the light of day until integration-time, occurs as late as possible and changes as little as possible.

> means to receive the common/pre-determined knowledge.

See the example provided. I may have been adding that example as an edit while you were responding. The map between exceptions and status codes is one example of common knowledge that must be shared between JSON and HTML message formats. Another is human-readable text representation of the exception. Another is differentiation and representation of sensitive information in a development mode.

**Comment by syndicatedshannon** [comment]
I'm still trying to sort this out, so I can perhaps validate these thoughts and put them in appropriate distinct issues, but it's rather time consuming to sort through disparate implementations in the existing code. I'm finding capabilities that I've not seen documented, so it also is taking some time to verify some of these capabilities I believe are not accessible, are actually not.

Discovering the ability to set `IStatusCodePagesFeature.Enabled` = false was a surprise to me. That smells strange to me, that another implementation might be expected to disable a specific feature to keep from breaking. This further complicates understanding possible interactions between middlewares that inspect and alter StatusCode and ResponseStarted, handle and rethrow exceptions, etc. The truth is hard (at least for me) to identify here.

Understanding the reason for difference between the implementation of page re-execute in the following locations: https://github.com/aspnet/AspNetCore/blob/c0161c7e777fffde7d914abbe0cd2716b6304dd7/src/Middleware/Diagnostics/src/ExceptionHandler/ExceptionHandlerMiddleware.cs#L117
https://github.com/aspnet/AspNetCore/blob/10863cfaead78e02eb788611fc1f64a889db6941/src/Middleware/Diagnostics/src/StatusCodePage/StatusCodePagesExtensions.cs#L202
https://github.com/aspnet/AspNetCore/blob/10863cfaead78e02eb788611fc1f64a889db6941/src/Middleware/Diagnostics/src/DeveloperExceptionPage/DeveloperExceptionPageMiddleware.cs#L178

The use of "Features" service locators to pass various values around the diagnostic middleware is taking me a while to sort out.

I appreciate that these components have grown organically. They generally work independently. But if I can't predict how they interact with other middleware, then IMO they aren't complete. I think their presence implies they are valuable in the library. I certainly want them there. I just can't currently use them in part or whole.

**Comment by IanKemp** [comment]
To paraphrase what I believe the gist of @syndicatedshannon's comments to be: while the work done for 3.0 (to return plaintext exception responses where appropriate) is great and solves 99% of developer gripes with `DeveloperExceptionPageMiddleware`, it's still not good enough because there's a lack of actual extensibility there.

@khellang's `ProblemDetails` package isn't an acceptable solution for me because it requires you to write your code in a way that's dependent on that library. I just want to be able to control the way that `DeveloperExceptionPageMiddleware` behaves when an exception occurs:

* get hold of the exception context (including stack trace information)
* decide in what format to respond (based on e.g. request headers)
* use a standard MVC view to return the relevant data

An example desired API in `Configure` in `Startup.cs`:

```csharp
app.UseDeveloperExceptionPage((request, exceptionContext) =>
{
    if (request.ContentType.IsText())
    {
        return View("MyPlainTextView", exceptionContext);
    }

    return View("MyHtmlView", exceptionContext);
});
```

**Comment by khellang** [comment]
@IanKemp That's exactly what the error handling middleware does. It let's you re-execute a route (i.e. an MVC action) and return whatever you want. That includes conneg.

**Comment by syndicatedshannon** [comment]
Thanks Ian. I'm still investigating. I know it's hard to pin down what my issue is, other than that I dislike the current implementation.

I'm working on an implementation of my request, primarily to better understand the problem, but possibly also to contribute. I've put more time and focus into this than I can afford.

If I were to bail on this investigation now, here are some of the things I would probably ask for:

- A clarified pattern or feature to divert/re-execute a response to a strongly-typed view. This is essential to error handling on "Accept: text/html" requests, and thus is implemented multiple times in existing error-handling code... none of which is publicly accessible.
- Reusable implementation of the existing developer Exception/StackTrace HTML visualizer, e.g. as an HTML partial with public supporting models. The current implementation is sweet, too bad I have to engage in severe CopyPasta.
- Single hub middleware that can be configured (e.g. via delegate injection) to perform most error handling steps (exception capture, error sanitization, status code assignment, response formatting) in a conventional manner. Currently I have to use 4 middlewares to get all of this in multiple response formats, and it's really confusing when some middlewares "bow out" because hidden flags have been set.
- Support for reusable error responses provided a strongly-typed model between Exceptional and non-Exceptional errors. Exception handling is important, but ideally we handle predictable problems via a non-exceptional workflow; chances are we'll want to use some of the same code either way. We can't do that easily if we have to carve up our handlers along the current inconsistent axes of sometimes Development/Production, sometimes Status Code, sometimes JSON, etc.

**Comment by syndicatedshannon** [comment]
> @IanKemp That's exactly what the error handling middleware does. It let's you re-execute a route (i.e. an MVC action) and return whatever you want. That includes conneg.

@khellang How would you call this from your own middleware? Keep in mind part of the issue is the assumption we'll engage custom middleware per comment by @davidfowl:

> I don’t think we’d build something that supports all of those features. It does sound like something that could be written in a custom middleware though. I’d suggest doing so and we’ll add any supporting platform features you’d need to achieve that.

**Comment by khellang** [comment]
> How would you call this from your own middleware?

You don't. Why would you need your own middleware for that? The error handling middleware captures the exception and re-execute the pipeline. Whether you have a custom middleware or an MVC controller to handle that exception, that's up to you.

**Comment by syndicatedshannon** [comment]
> You don't. Why would you need your own middleware for that?

I think we aren't talking the same language. Or maybe you are being intentionally adversarial, I'm actually not sure. But I'd rather not get in a chest-beating contest about the "right" implementation. That's why I referred to David's suggestion that we might need to build our own middleware. Additionally, I'd like to note you also built your own middleware for a related problem, and are suggesting others use it.

Again, I really don't want to get in an argument about it. I am often wrong, and may be here again, but, here are a few more specific examples of where I imagine such custom middleware may be necessary:

 - any time we need to provide something to handler, e.g. via DI, but the existing ExceptionHandlerOptions of PathString and RequestDelegate are insufficient
 - if we desire our view to receive a custom type
 - if we desire the outcome to be variant

**Comment by khellang** [comment]
I'm just getting more and more confused. I've tried to go back and read your posts multiple times, but it's very unclear what you're asking for (or even just talking about). Yes, I wrote my own middleware, it solves a bunch of people's issues. I thought it might help someone here as well. I don't know what you think I'm trying to accomplish here, but I'm out, sorry 🤷🏻‍♂️

**Comment by syndicatedshannon** [comment]
@khellang No worries. Thank you for your contribution.
